isolation/systemguard.cpp

```diff
@@ -0,0 +1,256 @@
+ * pdbg_target_traverse, recursion will exit when the method return 1
+ * else continues till the target is found
+ *
+ * @param[in] target - pdbg target to compare
+ * @param[inout] priv - data structure passed to the callback method
+ *
+ * @return 1 when target is found else 0
+ */
+int getGuardedTarget(struct pdbg_target* target, void* priv)
+{
+    GuardedTarget* guardTarget = reinterpret_cast<GuardedTarget*>(priv);
+    ATTR_PHYS_DEV_PATH_Type phyPath;
+    if (!DT_GET_PROP(ATTR_PHYS_DEV_PATH, target, phyPath))
+    {
+        if (strcmp(phyPath, guardTarget->phyDevPath.c_str()) == 0)
+        {
+            guardTarget->target = target;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/**
+ *
+ * @brief Get the location code of the target to be guarded
+ *
+ * @param[in] trgt - a pdbg target
+ *
+ * @return A string containing the FRU location code of the given chip. An empty
+ *         string indicates the target was null or the attribute does not exist
+ *         for this target.
+ */
+std::string getLocationCode(pdbg_target* trgt)
+{
+    if (nullptr == trgt)
+    {
+        return std::string{};
+    }
+    ATTR_LOCATION_CODE_Type val;
+    if (DT_GET_PROP(ATTR_LOCATION_CODE, trgt, val))
+    {
+        // Get the immediate parent in the devtree path and try again.
+        return getLocationCode(pdbg_target_parent(nullptr, trgt));
+    }
+    return val;
+}
+
+/**
+ * @brief convert input physical path to device tree understandable format
+ *
+ * @param[in] input - physical path of target to be guarded
+ *
+ * @return Device tree understandable physical path
+ **/
+std::string getDevTreePhyPathFormat(std::string_view input)
+{
+    constexpr std::string_view prefix = "physical:";
+
+    // Convert input to lowercase using std::ranges::transform
+    std::string lowercase_input(input);
+    std::ranges::transform(lowercase_input, lowercase_input.begin(),
+                           [](unsigned char c) { return std::tolower(c); });
+
+    // Check if the input starts with "physical:"
+    std::string result;
+    if (!lowercase_input.starts_with(prefix))
+    {
+        result = std::string(prefix) + lowercase_input;
+    }
+    else
+    {
+        result = std::string(lowercase_input);
+    }
+
+    // Check if the result starts with "physical:/"
+    constexpr std::string_view unwanted_slash = "physical:/";
+    if (result.starts_with(unwanted_slash))
+    {
+        result.erase(prefix.size(), 1); // Remove the extra slash
+    }
+
+    return result;
+}
+
+/**
+ * @brief Create a PEL with system guard
+ *
+ * @param[in] guardedTarget - a structure containing the info of the target to
+ *be guarded
+ * @param[in] sev - severity of the guard
+ *
+ **/
+void createPELWithSystemGuard(struct GuardedTarget& guardedTarget,
+                              const std::string sev)
+{
+    nlohmann::json pelJson;
+    nlohmann::json pelJsonArr = nlohmann::json::array();
```

Comment:
declare variable near the place where it is used in this case it is used at line 160